[opentitantool] Introduce binary protocol for HyperDebug gpio monitoring

jesultra opened the pull request:
HyperDebug supports logic analyzer functionality, in which it will record events on a given set of gpio pins. 
 `opentitantool` can then later be used to retrieve a transcript of every level change with microsecond timestamp.

This has been used by the GSC team to verify the reaction time of firmware under test.  Such testing involve typically a few handfuls of events, which can easily be transmitted via the textual protocol. However, we now plan on using the functionality for cases with 30000 events to be retrieved, which would take many tens of seconds to inefficently transmit via the console (which runs slow enough that the physical UART can keep up).

To improve performance, this CL introduces another Google-specific extension to the binary CMSIS-DAP protocol, for GPIO operations, and adds code to replicate the `gpio monitoring read` functionality.  (Starting and stopping the monitoring can still only be done through the textual protocol, those do not carry a large amount of data.  Though there may be a 80-character limit on a single command, which could impact the ability to monitor 5 or more signals at once, so in the future we may want to allow starting monitoring also through the binary protocol.)

Corresponding HyperDebug functionality implemented here:
https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/5128914

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -145,14 +148,54 @@ impl GpioPin for HyperdebugGpioPin {
+/// Size in bytes of the `RspGpioMonitoringHeader` struct, excluding the data field
+const GPIO_MONITORIN_HEADER_SIZE: usize = size_of::<RspGpioMonitoringHeader>() - USB_MAX_SIZE;
+
+#[derive(AsBytes, FromBytes, FromZeroes, Debug)]
+#[repr(packed)]
+struct RspGpioMonitoringHeader {
+    encapsulation_header: u8,
+    struct_size: u16,
+    status: u16,
+    start_levels: u16,
+    transcript_size: u16,
+    start_timestamp: u64,
+    end_timestamp: u64,
+    data: [u8; USB_MAX_SIZE],
+}
```
High-level overview of the struct: the encpasulation_header should not be part of this struct and should be handled separately. This would also remove the need of `repr(packed)` (would be `repr(C)`) since it'll then be naturally aligned. Data should not be part of this header.

This also needs doc comments detailing purpose of each field.

jesultra replied:
Done

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -145,14 +148,54 @@ impl GpioPin for HyperdebugGpioPin {
+    end_timestamp: u64,
+    data: [u8; USB_MAX_SIZE],
+}
+impl RspGpioMonitoringHeader {
+    fn new() -> Self {
+        Self {
+            encapsulation_header: 0,
+            struct_size: 0,
+            status: 0,
+            start_levels: 0,
+            start_timestamp: 0,
+            end_timestamp: 0,
+            transcript_size: 0,
+            data: [0u8; USB_MAX_SIZE],
+        }
+    }
+}
```
Please avoid using sentinel values.

jesultra replied:
Done

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+
+            let mut resp = RspGpioMonitoringHeader::new();
+            let mut bytecount = 0;
+            while bytecount < GPIO_MONITORIN_HEADER_SIZE {
+                let read_count = self.inner.usb_device.borrow().read_bulk(
+                    cmsis_interface.in_endpoint,
+                    &mut resp.as_bytes_mut()[bytecount..][..64],
+                )?;
+                ensure!(
+                    read_count > 0,
+                    TransportError::CommunicationError("Truncated GPIO response".to_string())
+                );
+                bytecount += read_count;
+            }
```
Why is this a loop? Given USB bulk transfers are packetised, should this always complete in a single transfer?

pamaury commented:
Can you confirm our understanding of the USB protocol used here (it's hard to figure it out form the device code since I am not familiar with the hyperdebug code): the device will send one or more USB bulk packets (full speed so up to 64 bytes each). The first packet contains the header and some data, and the remaining packets contain the rest of the data?
Also, will the device send a ZLP if the total transfer length is a multiple of 64 bytes, or not? (I am assuming not since your code treats a ZLP as an error).

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            }
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
+
+            let mut databytes: Vec<u8> = Vec::new();
```
```suggestion
            let mut data_bytes = Vec::with_capacity(resp.transcript_size as usize);
```

jesultra replied:
I have reworked the logic to do just a single `resize()`.

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
+
+            let mut databytes: Vec<u8> = Vec::new();
+            databytes.extend_from_slice(&resp.data[..bytecount - GPIO_MONITORIN_HEADER_SIZE]);
```
You should just skip bytes here and not using it as starting index or loop condition later.

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+
+            while databytes.len() < skip_bytes + resp.transcript_size as usize {
+                let original_length = databytes.len();
+                databytes.resize(original_length + 64, 0u8);
```
Instead of growing and shrinking, perhaps resize to full expected size ahead of time and just keep filling it.

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                // since the previous event (on any signal, not necessarily on that same one).
+                cur_time += value >> signal_bits;
+                events.push(MonitoringEvent {
+                    signal_index: (value & signal_mask) as u8,
```
`(value & signal_mask)` is used many times. Perhaps save it to a variable first.

jesultra replied:
Done

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                events.push(MonitoringEvent {
+                    signal_index: (value & signal_mask) as u8,
+                    edge: if cur_level[(value & signal_mask) as usize] {
+                        cur_level[(value & signal_mask) as usize] = false;
```
This could be a `cur_level[index] ^= true;` after the push, or, `cur_level ^= 1 << index;` if you opted to use bitmask instead of creating an array.

jesultra replied:
Done

nbdd0121 reviewed (changes requested):
The lowRISC software team reviewed this pull request together in a review session.

We have some detailed suggestions below, but also at the high level have some questions about how this is implemented.

Firstly this is missing comments. Please include comments detailing the protocol so that code readers do not have to get it from another source. If there is a canonical description of the protocol, please also include a link.

We are skeptical about the way the struct is organised. It's named header but it is clearly including more things than just the header, i.e. it contains a protocol identification byte and data bytes. The data bytes also do not reflect the actual maximum size, and if we understand correctly that's only for ensuring that USB bulk reads are not overshort.

Another thing is that this is a wire protocol, it needs to be conscious of the byte order. You can either use zerocopy's byte order primitives, or perhaps just parse the bytes using byteorder crate directly without creating a struct. Given the small number of fields it might even aid readability. Although, given that we don't really support BE machines, feel free to ignore this suggestion.

PS: In https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/5128914, gpio.c L1168, it converts `tx_buffer + 1` (with alignment 1) to `struct gpio_monitoring_header_t` (with alignment 8) and is undefined behaviour.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                .borrow()
+                .write_bulk(cmsis_interface.out_endpoint, &pkt)?;
+
+            let mut resp = RspGpioMonitoringHeader::new();
```
Done

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                    "Unrecognized CMSIS-DAP response to GPIO request".to_string()
+                )
+            );
+            match resp.status {
+                0 => (),
+                n => bail!(TransportError::CommunicationError(format!(
+                    "HyperDebug error: {}",
+                    n
+                ))),
+            }
```
It just happens that I added handling for one specific error code, which deserves a descriptive error message, since it can be triggered by the operator, and not only by bugs in opentitantool/HyperDebug.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                    n
+                ))),
+            }
+            let skip_bytes = resp.struct_size as usize - (GPIO_MONITORIN_HEADER_SIZE - 1);
```
I have addressed both points.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            // Figure out how many of the low bits are used for storing the index of the signal
+            // hanving changed. (If only one signal, no bits are used, if two signals, then one
+            // bit is used, if three or four, then two bits are used, etc.)
+            let mut signal_bits = 0;
+            while (pin_names.len() - 1) >> signal_bits != 0 {
+                signal_bits += 1;
+            }
```
I chose the latter of the two suggestions, since I think the name of the method `leading_zeros()` explains everything to any casual reader, who like me may not be familiar with either `ilog2()` or `leading_zeros()`.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            let signal_mask = (1u64 << signal_bits) - 1;
+
+            let mut cur_time: u64 = resp.start_timestamp;
+            let mut cur_level = Vec::new();
```
Yes, I chose `Vec<bool>` because I thought that the code in the loop would become cleaner, but it seems that it is causing more trouble than it is saving.  I have changed the code to use a bitmask in the format of `start_levels`.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,6 +263,121 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            // Now decode the list of events, each consisting of a variable legnth encoded 64-bit
+            // integer.
+            while idx < databytes.len() {
+                // Read 7 bits from each byte, least significant byte first.  High bit of one
+                // indicates more bytes belong to the same value.
+                let mut value = 0u64;
+                let mut shift = 0;
+                loop {
+                    value |= ((databytes[idx] & 0x7F) as u64) << shift;
+                    shift += 7;
+                    idx += 1;
+                    if (databytes[idx - 1] & 0x80) == 0 {
+                        break;
+                    }
+                }
```
I was not aware that this was a standard format, with an existing crate for decoding.  Rather than add to dependencies, I have factored out the leb128 decoding, and used that name, so that readers can immediately recognize the name and format.

jesultra commented:
Thank you for the very detailed review.

> Can you confirm our understanding of the USB protocol used here (it's hard to figure it out form the device code since I am not familiar with the hyperdebug code): the device will send one or more USB bulk packets (full speed so up to 64 bytes each). The first packet contains the header and some data, and the remaining packets contain the rest of the data? Also, will the device send a ZLP if the total transfer length is a multiple of 64 bytes, or not? (I am assuming not since your code treats a ZLP as an error).

Yes, on the HyperDebug firmware side, I use a USB-queueing library also used with e.g. the UART USB interfaces.  This library does not deal with "transfers" terminated with a short (possibly zero length) packet, but thinks of the data a a single stream, and may in theory send any number of bytes in USB packets, though as I enqueue bytes very quickly in this case, unlike a physical UART, they will practically always arrive in 64-byte packets except the last one.  My code here is still prepared for short packets "in the middle" of the response, though.

In the future, I may choose to change the HyperDebug firmware to provide the usual transmission boundaries, if we find that desirable.  I chose the above mostly because the USB streaming code library was there already, and is well tested, allowing me to spend time on the higher level functions.

> Firstly this is missing comments. Please include comments detailing the protocol so that code readers do not have to get it from another source. If there is a canonical description of the protocol, please also include a link.

I do not have a document describing the protocol, only comments in the HyperDebug source code.  I have added more comments in this file, and a reference to the HyperDebug file.

> We are skeptical about the way the struct is organised. It's named header but it is clearly including more things than just the header, i.e. it contains a protocol identification byte and data bytes. The data bytes also do not reflect the actual maximum size, and if we understand correctly that's only for ensuring that USB bulk reads are not overshort.

Yes, I could not do `read_bulk(size_of::<...>)` without losing the remaning bytes from the first 64-byte packet being sent, which is why I added the extra `data` field to the header struct.  Using `FromBytes` as you suggested is much cleaner.

> Another thing is that this is a wire protocol, it needs to be conscious of the byte order. You can either use zerocopy's byte order primitives, or perhaps just parse the bytes using byteorder crate directly without creating a struct. Given the small number of fields it might even aid readability. Although, given that we don't really support BE machines, feel free to ignore this suggestion.

Yes, I used to be much more conscious of wire protocols, preferring big endian on the wires.  But the EC codebase makes no such attempt, at it seems that BE machines are not common anymore, so I think I will ignore this issue.

nbdd0121 reviewed (commented):
Thanks for updating. This looks much better now.

pamaury commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -220,13 +273,127 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+                vec![0u8; 1 + size_of::<RspGpioMonitoringHeader>() + USB_MAX_SIZE];
+            let mut bytecount = 0;
+
+            while bytecount < 1 + size_of::<RspGpioMonitoringHeader>() {
```
The header always fits in one packet. Would it make sense to change the code to first do an initial `read_bulk` to get the header + part of the data. With this, we now know the size of the full data so we can allocate a buffer for it and read it with several `read_bulk`?

nbdd0121 replied:
From my interpretation of @jesultra's comment it sounds like they're using USB as a stream and doesn't guarantee the header'll be sent in a single packet?

pamaury replied:
Ah sorry, I missed @jesultra 's comment on the USB protocol.

jesultra commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -268,7 +442,22 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+
+/// Read 7 bits from each byte, least significant byte first.  High bit of one indicates more
+/// bytes belong to the same value.
+fn decode_leb128(idx: &mut usize, databytes: &[u8]) -> u64 {
```
I have modified this routine to return `Result<u64>`, and for good measure made it not mutate `idx` in the case of errors.

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -268,7 +442,29 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+    let mut shift = 0;
+    while i < databytes.len() {
+        let byte = databytes[i];
+        value |= ((byte & 0x7F) as u64) << shift;
```
This may overflow there are 10 bytes without MSB set.

jesultra replied:
I have inserted a check to return an error in case of a tenth byte in the encoding, meaning that the routine accepts at most 63 bits of data for an u64.  I do not think it is worth supporting the most significant 64th bit being set through the lowest bit in the tenth byte of encoding, as it would make error handling more complicated.

nbdd0121 reviewed (approved):
LGTM apart a final nit

nbdd0121 commented on sw/host/opentitanlib/src/transport/hyperdebug/gpio.rs:
```diff
@@ -268,7 +442,35 @@ impl GpioMonitoring for HyperdebugGpioMonitoring {
+            *idx = i;
+            return Ok(value);
+        }
+        if shift + 7 > 64 {
```
shift is already incremented by 7 above?

nbdd0121 replied:
Ah, this is to error when `shift == 63`

jesultra replied:
Yes, but even having a shift value of e.g. 60 for next round would allow an overflow, as a value of 0x7F shifted 60 bits would need 67 bits of storage.  That is why this line ensures that you can shift any 7-bit value by `shift` bits, and it still does not exceed 64 bits.

pamaury commented on sw/host/opentitanlib/src/transport/hyperdebug/mod.rs:
```diff
@@ -331,6 +344,41 @@ impl<T: Flavor> Hyperdebug<T> {
+            .read_bulk(cmsis_interface.in_endpoint, &mut resp)?;
+        let resp = &resp[..bytecount];
+        ensure!(
+            bytecount >= 4 && resp[0] == Self::CMSIS_DAP_CUSTOM_COMMAND_GOOGLE_INFO && resp[1] >= 2,
```
maybe document what the `resp[1] >= 2` is? I assume that this is the number of bytes that follow the header?

jesultra replied:
Done
